v0.10.0-rc.0 stacktrace when during cluster deletion

/kind bug

**What steps did you take and what happened:**
During cluster deletion my CAPO controller-manager crashes with the following stacktrace:

```log
I0412 11:45:50.916394       1 openstackcluster_controller.go:260] "Deleting Bastion" controller="openstackcluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="OpenStackCluster" OpenStackCluster="default/hux-lab1" namespace="default" name="hux-lab1" reconcileID="2b4f4d91-fc50-4a8a-b2f8-2cb3db96d50c" cluster="hux-lab1"
I0412 11:45:51.204917       1 controller.go:115] "Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference" controller="openstackcluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="OpenStackCluster" OpenStackCluster="default/hux-lab1" namespace="default" name="hux-lab1" reconcileID="2b4f4d91-fc50-4a8a-b2f8-2cb3db96d50c"
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x60 pc=0x1b912b0]

goroutine 425 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:116 +0x1e5
panic({0x1db33e0?, 0x3646f90?})
	/usr/local/go/src/runtime/panic.go:770 +0x132
sigs.k8s.io/cluster-api-provider-openstack/controllers.bastionToInstanceSpec(0xc000a56f08, 0xc000c12820?)
	/workspace/controllers/openstackcluster_controller.go:563 +0xb0
sigs.k8s.io/cluster-api-provider-openstack/controllers.deleteBastion(0xc0009fbd70, 0xc000c12820, 0xc000a56f08)
	/workspace/controllers/openstackcluster_controller.go:297 +0x2db
sigs.k8s.io/cluster-api-provider-openstack/controllers.(*OpenStackClusterReconciler).reconcileDelete(0xc000547740, {0x242d758, 0xc000e81770}, 0xc0009fbd70, 0xc000c12820, 0xc000a56f08)
	/workspace/controllers/openstackcluster_controller.go:160 +0x2b2
sigs.k8s.io/cluster-api-provider-openstack/controllers.(*OpenStackClusterReconciler).Reconcile(0xc000547740, {0x242d758, 0xc000e81770}, {{{0xc000b46040?, 0x0?}, {0xc000b46088?, 0xc00089dd50?}}})
	/workspace/controllers/openstackcluster_controller.go:126 +0x6cc
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x2432cc8?, {0x242d758?, 0xc000e81770?}, {{{0xc000b46040?, 0xb?}, {0xc000b46088?, 0x0?}}})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:119 +0xb7
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0006faaa0, {0x242d790, 0xc000575b80}, {0x1e7dc40, 0xc000329400})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:316 +0x3bc
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0006faaa0, {0x242d790, 0xc000575b80})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:266 +0x1c9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:227 +0x79
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2 in goroutine 200
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:223 +0x50c
```

openstackcluster:
```yaml
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: OpenStackCluster
metadata:
  creationTimestamp: "2024-04-12T11:12:54Z"
  deletionGracePeriodSeconds: 0
  deletionTimestamp: "2024-04-12T11:43:04Z"
  finalizers:
  - openstackcluster.infrastructure.cluster.x-k8s.io
  generation: 4
  labels:
    cluster.x-k8s.io/cluster-name: hux-lab1
  name: hux-lab1
  namespace: default
  ownerReferences:
  - apiVersion: cluster.x-k8s.io/v1beta1
    blockOwnerDeletion: true
    controller: true
    kind: Cluster
    name: hux-lab1
    uid: 28be47df-1ba5-417a-81f5-84916aca2a82
  resourceVersion: "10137"
  uid: 38e38891-bbe4-4e45-9161-129ce22ad71c
spec:
  apiServerLoadBalancer:
    additionalPorts:
    - [REDACTED]
    allowedCIDRs:
    - [REDACTED]
    enabled: true
  controlPlaneAvailabilityZones:
  - sto1
  - sto2
  - sto3
  controlPlaneEndpoint:
    host: [REDACTED]
    port: 6443
  externalNetwork:
    id: 600b8501-78cb-4155-9c9f-23dfcba88828
  identityRef:
    cloudName: elastx
    name: hux-lab1-cloud-config
  managedSecurityGroups:
    allNodesSecurityGroupRules:
    - description: Allow BGP traffic
      direction: ingress
      etherType: IPv4
      name: BGP (Calico)
      portRangeMax: 179
      portRangeMin: 179
      protocol: tcp
      remoteManagedGroups:
      - controlplane
      - worker
    - description: Allow IP-in-IP traffic
      direction: ingress
      etherType: IPv4
      name: IP-in-IP (calico)
      protocol: "4"
      remoteManagedGroups:
      - controlplane
      - worker
    allowAllInClusterTraffic: true
  managedSubnets:
  - cidr: 10.128.0.0/22
    dnsNameservers:
    - [REDACTED]
status:
  apiServerLoadBalancer:
    allowedCIDRs:
    - [REDACTED]
    id: e01a7390-f43c-4c5f-b693-e3a42a1bdbb5
    internalIP: 10.128.2.205
    ip: [REDACTED]
    loadBalancerNetwork:
      id: ""
      name: ""
    name: k8s-clusterapi-cluster-default-hux-lab1-kubeapi
  controlPlaneSecurityGroup:
    id: bb9e434c-64f4-40e0-8a0d-d0ba6ddfde71
    name: k8s-cluster-default-hux-lab1-secgroup-controlplane
  externalNetwork:
    id: 600b8501-78cb-4155-9c9f-23dfcba88828
    name: elx-public1
  failureDomains:
    sto1:
      controlPlane: true
    sto2:
      controlPlane: true
    sto3:
      controlPlane: true
  network:
    id: ae6b5ea2-a851-4c09-a12b-9623ffae0da0
    name: k8s-clusterapi-cluster-default-hux-lab1
    subnets:
    - cidr: 10.128.0.0/22
      id: f0bfc3d3-26e6-4fec-a8be-e8f5a3dd7179
      name: k8s-clusterapi-cluster-default-hux-lab1
  ready: true
  router:
    id: 296b2430-afd8-49d9-b42f-b2edfe833927
    ips:
    - [REDACTED]
    name: k8s-clusterapi-cluster-default-hux-lab1
  workerSecurityGroup:
    id: 162ad19c-b8ea-4935-8510-50eac441839d
    name: k8s-cluster-default-hux-lab1-secgroup-worker
```


**What did you expect to happen:**
The cluster to be deleted


**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]


**Environment:**

- Cluster API Provider OpenStack version (Or `git rev-parse HEAD` if manually built): 0.10.0-rc.0
- Cluster-API version: [v1.6.3](https://github.com/kubernetes-sigs/cluster-api/releases/tag/v1.6.3)
- OpenStack version: Ussuri
- Minikube/KIND version: kind [v0.22.0](https://github.com/kubernetes-sigs/kind/releases/tag/v0.22.0)
- Kubernetes version (use `kubectl version`): 1.29.1
- OS (e.g. from `/etc/os-release`): Ubuntu 22.04

--- a/Makefile
+++ b/Makefile
@@ -161,7 +161,8 @@ e2e-templates: $(addprefix $(E2E_NO_ARTIFACT_TEMPLATES_DIR)/, \
 		 cluster-template.yaml \
 		 cluster-template-flatcar.yaml \
                  cluster-template-k8s-upgrade.yaml \
-		 cluster-template-flatcar-sysext.yaml)
+		 cluster-template-flatcar-sysext.yaml \
+		 cluster-template-no-bastion.yaml)
 # Currently no templates that require CI artifacts
 # $(addprefix $(E2E_TEMPLATES_DIR)/, add-templates-here.yaml) \
 

--- a/controllers/openstackcluster_controller.go
+++ b/controllers/openstackcluster_controller.go
@@ -291,15 +291,13 @@ func deleteBastion(scope *scope.WithLogger, cluster *clusterv1.Cluster, openStac
 	}
 
 	// If no instance was created we currently need to check for orphaned
-	// volumes. This requires resolving the instance spec.
-	// TODO: write volumes to status resources on creation so this is no longer required.
+	// volumes.
 	if instanceStatus == nil {
-		instanceSpec, err := bastionToInstanceSpec(openStackCluster, cluster)
-		if err != nil {
-			return err
-		}
-		if err := computeService.DeleteVolumes(instanceSpec); err != nil {
-			return fmt.Errorf("delete volumes: %w", err)
+		bastion := openStackCluster.Spec.Bastion
+		if bastion != nil && bastion.Spec != nil {
+			if err := computeService.DeleteVolumes(bastionName(cluster.Name), bastion.Spec.RootVolume, bastion.Spec.AdditionalBlockDevices); err != nil {
+				return fmt.Errorf("delete volumes: %w", err)
+			}
 		}
 	} else {
 		instanceNS, err := instanceStatus.NetworkStatus()

--- a/controllers/openstackmachine_controller.go
+++ b/controllers/openstackmachine_controller.go
@@ -300,19 +300,12 @@ func (r *OpenStackMachineReconciler) reconcileDelete(scope *scope.WithLogger, cl
 	}
 
 	// If no instance was created we currently need to check for orphaned
-	// volumes. This requires resolving the instance spec.
-	// TODO: write volumes to status resources on creation so this is no longer required.
-	if instanceStatus == nil && openStackMachine.Status.Resolved != nil {
-		instanceSpec, err := machineToInstanceSpec(openStackCluster, machine, openStackMachine, "")
-		if err != nil {
-			return ctrl.Result{}, err
-		}
-		if err := computeService.DeleteVolumes(instanceSpec); err != nil {
+	// volumes.
+	if instanceStatus == nil {
+		if err := computeService.DeleteVolumes(openStackMachine.Name, openStackMachine.Spec.RootVolume, openStackMachine.Spec.AdditionalBlockDevices); err != nil {
 			return ctrl.Result{}, fmt.Errorf("delete volumes: %w", err)
 		}
-	}
-
-	if instanceStatus != nil {
+	} else {
 		if err := computeService.DeleteInstance(openStackMachine, instanceStatus); err != nil {
 			conditions.MarkFalse(openStackMachine, infrav1.InstanceReadyCondition, infrav1.InstanceDeleteFailedReason, clusterv1.ConditionSeverityError, "Deleting instance failed: %v", err)
 			return ctrl.Result{}, fmt.Errorf("delete instance: %w", err)

--- a/pkg/cloud/services/compute/instance.go
+++ b/pkg/cloud/services/compute/instance.go
@@ -431,11 +431,14 @@ func (s *Service) DeleteInstance(eventObject runtime.Object, instanceStatus *Ins
 }
 
 // DeleteVolumes deletes any cinder volumes which were created for the instance.
-// Note that this must only be called when the server was not successfully
+// Note that this need only be called when the server was not successfully
 // created. If the server was created the volume will have been added with
 // DeleteOnTermination=true, and will be automatically cleaned up with the
 // server.
-func (s *Service) DeleteVolumes(instanceSpec *InstanceSpec) error {
+// We don't pass InstanceSpec here because we only require instance name,
+// rootVolume, and additionalBlockDevices, and resolving the whole InstanceSpec
+// introduces unnecessary failure modes.
+func (s *Service) DeleteVolumes(instanceName string, rootVolume *infrav1.RootVolume, additionalBlockDevices []infrav1.AdditionalBlockDevice) error {
 	/*
 		Attaching volumes to an instance is a two-step process:
 
@@ -455,13 +458,13 @@ func (s *Service) DeleteVolumes(instanceSpec *InstanceSpec) error {
 		DeleteOnTermination will ensure it is deleted in that case.
 	*/
 
-	if hasRootVolume(instanceSpec) {
-		if err := s.deleteVolume(instanceSpec.Name, "root"); err != nil {
+	if rootVolume != nil && rootVolume.SizeGiB > 0 {
+		if err := s.deleteVolume(instanceName, "root"); err != nil {
 			return err
 		}
 	}
-	for _, volumeSpec := range instanceSpec.AdditionalBlockDevices {
-		if err := s.deleteVolume(instanceSpec.Name, volumeSpec.Name); err != nil {
+	for _, volumeSpec := range additionalBlockDevices {
+		if err := s.deleteVolume(instanceName, volumeSpec.Name); err != nil {
 			return err
 		}
 	}

--- a/test/e2e/data/kustomize/no-bastion/kustomization.yaml
+++ b/test/e2e/data/kustomize/no-bastion/kustomization.yaml
@@ -1,0 +1,11 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+
+resources:
+- ../default
+
+patches:
+- path: patch-no-bastion.yaml
+  target:
+    kind: OpenStackCluster
+    name: \${CLUSTER_NAME}

--- a/test/e2e/data/kustomize/no-bastion/patch-no-bastion.yaml
+++ b/test/e2e/data/kustomize/no-bastion/patch-no-bastion.yaml
@@ -1,0 +1,3 @@
+---
+- op: remove
+  path: /spec/bastion

--- a/test/e2e/shared/common.go
+++ b/test/e2e/shared/common.go
@@ -214,51 +214,55 @@ func (o OpenStackLogCollector) CollectMachineLog(ctx context.Context, management
 		return fmt.Errorf("error writing server JSON %s: %s", serverJSON, err)
 	}
 
-	srvUser := o.E2EContext.E2EConfig.GetVariable(SSHUserMachine)
-	executeCommands(
-		ctx,
-		o.E2EContext.Settings.ArtifactFolder,
-		o.E2EContext.Settings.Debug,
-		outputPath,
-		ip,
-		openStackCluster.Status.Bastion.FloatingIP,
-		srvUser,
-		[]command{
-			// don't do this for now, it just takes to long
-			// {
-			//	title: "systemd",
-			//	cmd:   "journalctl --no-pager --output=short-precise | grep -v  'audit:\\|audit\\['",
-			// },
-			{
-				title: "kern",
-				cmd:   "journalctl --no-pager --output=short-precise -k",
+	if openStackCluster.Status.Bastion == nil {
+		Logf("Skipping log collection for machine %q since no bastion is available", m.Name)
+	} else {
+		srvUser := o.E2EContext.E2EConfig.GetVariable(SSHUserMachine)
+		executeCommands(
+			ctx,
+			o.E2EContext.Settings.ArtifactFolder,
+			o.E2EContext.Settings.Debug,
+			outputPath,
+			ip,
+			openStackCluster.Status.Bastion.FloatingIP,
+			srvUser,
+			[]command{
+				// don't do this for now, it just takes to long
+				// {
+				//	title: "systemd",
+				//	cmd:   "journalctl --no-pager --output=short-precise | grep -v  'audit:\\|audit\\['",
+				// },
+				{
+					title: "kern",
+					cmd:   "journalctl --no-pager --output=short-precise -k",
+				},
+				{
+					title: "containerd-info",
+					cmd:   "crictl --runtime-endpoint unix:///run/containerd/containerd.sock info",
+				},
+				{
+					title: "containerd-containers",
+					cmd:   "crictl --runtime-endpoint unix:///run/containerd/containerd.sock ps",
+				},
+				{
+					title: "containerd-pods",
+					cmd:   "crictl --runtime-endpoint unix:///run/containerd/containerd.sock pods",
+				},
+				{
+					title: "cloud-final",
+					cmd:   "journalctl --no-pager -u cloud-final",
+				},
+				{
+					title: "kubelet",
+					cmd:   "journalctl --no-pager -u kubelet.service",
+				},
+				{
+					title: "containerd",
+					cmd:   "journalctl --no-pager -u containerd.service",
+				},
 			},
-			{
-				title: "containerd-info",
-				cmd:   "crictl --runtime-endpoint unix:///run/containerd/containerd.sock info",
-			},
-			{
-				title: "containerd-containers",
-				cmd:   "crictl --runtime-endpoint unix:///run/containerd/containerd.sock ps",
-			},
-			{
-				title: "containerd-pods",
-				cmd:   "crictl --runtime-endpoint unix:///run/containerd/containerd.sock pods",
-			},
-			{
-				title: "cloud-final",
-				cmd:   "journalctl --no-pager -u cloud-final",
-			},
-			{
-				title: "kubelet",
-				cmd:   "journalctl --no-pager -u kubelet.service",
-			},
-			{
-				title: "containerd",
-				cmd:   "journalctl --no-pager -u containerd.service",
-			},
-		},
-	)
+		)
+	}
 	return nil
 }
 

--- a/test/e2e/shared/defaults.go
+++ b/test/e2e/shared/defaults.go
@@ -46,6 +46,7 @@ const (
 	OpenStackNodeMachineFlavor = "OPENSTACK_NODE_MACHINE_FLAVOR"
 	SSHUserMachine             = "SSH_USER_MACHINE"
 	FlavorDefault              = ""
+	FlavorNoBastion            = "no-bastion"
 	FlavorWithoutLB            = "without-lb"
 	FlavorMultiNetwork         = "multi-network"
 	FlavorMultiAZ              = "multi-az"

--- a/test/e2e/suites/e2e/e2e_test.go
+++ b/test/e2e/suites/e2e/e2e_test.go
@@ -245,6 +245,42 @@ var _ = Describe("e2e tests [PR-Blocking]", func() {
 		})
 	})
 
+	Describe("Workload cluster (no bastion)", func() {
+		It("should be creatable and deletable", func() {
+			shared.Logf("Creating a cluster")
+			clusterName := fmt.Sprintf("cluster-%s", namespace.Name)
+			configCluster := defaultConfigCluster(clusterName, namespace.Name)
+			configCluster.ControlPlaneMachineCount = ptr.To(int64(1))
+			configCluster.WorkerMachineCount = ptr.To(int64(1))
+			configCluster.Flavor = shared.FlavorNoBastion
+			createCluster(ctx, configCluster, clusterResources)
+			md := clusterResources.MachineDeployments
+
+			workerMachines := framework.GetMachinesByMachineDeployments(ctx, framework.GetMachinesByMachineDeploymentsInput{
+				Lister:            e2eCtx.Environment.BootstrapClusterProxy.GetClient(),
+				ClusterName:       clusterName,
+				Namespace:         namespace.Name,
+				MachineDeployment: *md[0],
+			})
+			controlPlaneMachines := framework.GetControlPlaneMachinesByCluster(ctx, framework.GetControlPlaneMachinesByClusterInput{
+				Lister:      e2eCtx.Environment.BootstrapClusterProxy.GetClient(),
+				ClusterName: clusterName,
+				Namespace:   namespace.Name,
+			})
+			Expect(workerMachines).To(HaveLen(1))
+			Expect(controlPlaneMachines).To(HaveLen(1))
+
+			shared.Logf("Waiting for worker nodes to be in Running phase")
+			statusChecks := []framework.MachineStatusCheck{framework.MachinePhaseCheck(string(clusterv1.MachinePhaseRunning))}
+			machineStatusInput := framework.WaitForMachineStatusCheckInput{
+				Getter:       e2eCtx.Environment.BootstrapClusterProxy.GetClient(),
+				Machine:      &workerMachines[0],
+				StatusChecks: statusChecks,
+			}
+			framework.WaitForMachineStatusCheck(ctx, machineStatusInput, e2eCtx.E2EConfig.GetIntervals(specName, "wait-machine-status")...)
+		})
+	})
+
 	Describe("Workload cluster (flatcar)", func() {
 		It("should be creatable and deletable", func() {
 			// Flatcar default user is "core"

